### PR TITLE
Fix crashes in JFactoryPodioT::ClearData()

### PR DESCRIPTION
**Pull request by veprbl**
We are observing inconsistent crashes in delete operator invocation in ClearData() on certain events. These faults occur on macOS with a high rate, specifically for CalorimeterHitContribution. We also have reports of some percent of reconstruction jobs (on Linux) crashing specifically on ReconstructedParticle.

There might be an issue with PODIO object clones, they should be pointing to the same object data. There doesn't seem to be any facility for reference counting.

A slight complication with addressing this is that the failure comes in the implementation FactoryPodioT from JANA2 libraries. That factory is implicitly instantiated as a result of EICrecon calling the `JEvent::InsertCollectionAlreadyInFrame` method in `JEventSourcePODIO`. This PR switches to EICrecon variant of the same code to have a chance of handling the issue locally.

It is not clear if mData is used or just populated for compatibility. This is an attempt at disabling the functionality in order to mitigate the issue.

Resolves: #949
cc @wdconinc @rahmans1

### CODE REVIEW

File: src/services/io/podio/JEventSourcePODIO.cc [hunk 1]
```diff
@@ -9,6 +9,7 @@
 // it is limited to processing only a single event at a time.
```
Comment (veprbl) [inline src/services/io/podio/JEventSourcePODIO.cc:9]: These changes would need to be duplicated to the JEventSourceLegacyPODIO.cc

**Comment by nathanwbrei** [comment]
mData is populated for compatibility so that the old-style `JFactoryT<T>::Set()`, `JEvent::Get<T>()` work as expected. If you use `{Set,Get}Collection` everywhere instead you should be fine. 
However we should be trying to get rid of these hacked-up JANA facilities and upstream the long-term fixes.`mData` can't be the root cause of the heisenbug. PODIO shouldn't need reference counting because the underlying data is owned by the collection, not the lightweight object wrappers.

**Comment by nathanwbrei** [comment]
Are these failures all happening with subset collections? Could it be that PODIO is failing when deleting the object wrappers that point to an already-deleted collection?

**Comment by veprbl** [comment]
> mData is populated for compatibility so that the old-style `JFactoryT<T>::Set()`, `JEvent::Get<T>()` work as expected. If you use `{Set,Get}Collection` everywhere instead you should be fine.

That was my working assumption. Once CI runs again on this PR, we can see if this breaks anything (the `JEvent::Get` should be mostly phased out by now).

> However we should be trying to get rid of these hacked-up JANA facilities and upstream the long-term fixes.

Sure, here I'm going oppositely to #837.
Would be nice to be able to supply custom dependencies in our CI.

> `mData` can't be the root cause of the heisenbug.

It shouldn't be, yes.

> PODIO shouldn't need reference counting because the underlying data is owned by the collection, not the lightweight object wrappers.

This is good to know. I wonder if there can be some sort of lifetime issues if the input EDM4hep collections are directly saved into EDM4eic.

**Comment by veprbl** [comment]
Even though we don't use mData and don't plan to. Maybe, let's keep the "feature" by attempting the "unlink" strategy.

**Comment by veprbl** [comment]
JANA PR https://github.com/JeffersonLab/JANA2/pull/243

**Comment by veprbl** [comment]
@wdconinc Review?

**Comment by wdconinc** [comment]
Is this able to run with the LSAN enforcement we have in place in main?

**Review by wdconinc — APPROVED** [review]
Tested locally. No LSAN warnings emitted (I was just worried that PODIO wasn't going to do the deep clean that ClearData was doing before).

**Comment by veprbl** [comment]
> Is this able to run with the LSAN enforcement we have in place in main?

This is designed to not leak memory, at least in EICrecon (JANA, in general, is another story https://github.com/JeffersonLab/JANA2/pull/243#issuecomment-1745477936).

**Pull request merged**